File: week4/java/PartitionToKEqualSumSubsets_698.java
class Solution {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        
        if (k >= nums.length) return false;
        
        int max_num = 0;
        int target = 0;
        for (int num : nums) {
            target += num;
            max_num = Math.max(max_num, num);
        }
        
        if (target % k != 0 || target / k < max_num) return false;
        
        target /= k;
        boolean[] visited = new boolean[nums.length];
        return partition(nums, 0, 0, target, visited, k);
    }
    
    private boolean partition(int[] nums, int index, int cur_sum, int target, boolean[] visited, int k) {
        // base case: found all k partitions
        if (k == 0) return true;
        // current sum is not correct
        if (cur_sum > target) return false;
        // found 1 partition, continue
        if (cur_sum == target) return partition(nums, 0, 0, target, visited, k - 1);
        
        for (int i = index; i < nums.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                if (partition(nums, i + 1, cur_sum + nums[i], target, visited, k)) {
                    return true;
                }
                visited[i] = false;
            }
        }
        return false;
    }
}

File: week4/python/PartitionToKEqualSumSubsets_698.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        
#         sums = [0]*k
#         target = sum(nums) // k
#         nums.sort(reverse=True)
#         len_nums = len(nums)
        
#         def findPartition(index: int) -> bool:

#             if index == len_nums:
#                 return len(set(sums)) == 1
            
#             for count in range(k):
#                 sums[count] += nums[index]
#                 if sums[count] <= target and findPartition(index+1):
#                     return True
                
#                 sums[count] -= nums[index]
#                 if sums[count] == 0:
#                     break
                    
#             return False        
        
#         return findPartition(0)

        target = sum(nums)
        if target % k != 0:
            return False
        
        target //= k
        len_nums = len(nums)
        visited = [False] * len_nums


        def findPartition(index: int, cur_sum: int, found: int) -> bool:
            
            if found == k:
                return True
            
            if cur_sum == target:
                return findPartition(0, 0, found + 1)
            
            for i in range(index, len(nums)):
                if visited[i] or cur_sum + nums[i] > target:
                    continue
                
                visited[i] = True
                if findPartition(i + 1, cur_sum + nums[i], found):
                    return True
                
                visited[i] = False
            
            return False
        
        return findPartition(0, 0, 0)
